clsp/fusion_singleton.clsp

```diff
@@ -31,149 +26,98 @@
-                    (sha256tree (list offer_nonce (notarized_payment (f nft_next_puzzlehashes) )))
-                )
-            ) 
-            conditions_so_far
-          )
-        )
-        (c
-          (list ASSERT_PUZZLE_ANNOUNCEMENT
-              (sha256 
-                  (calculate_full_puzzle_hash SINGLETON_MOD_HASH (f (f nft_ids)) LAUNCHER_PUZZLE_HASH (f nft_inner_puzzlehashes))
-                  (sha256tree (list offer_nonce (notarized_payment (f nft_next_puzzlehashes) )))
-              )
-          ) 
-          conditions_so_far
-        )
+  ; maintain announcement loop with p2 for unlocking to puzzlehash
+  (defun-inline announce_p2_condition (SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (nft_launcher_id coin_id inner_puzzlehash target_inner_puzzlehash) singleton_coin_id release_puzzlehash)
+    (list CREATE_PUZZLE_ANNOUNCEMENT (sha256 singleton_coin_id nft_launcher_id release_puzzlehash))
+  )
+
+  ; maintain announcement loop with p2 for unlocking to puzzlehash
+  (defun-inline assert_p2_condition (SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (nft_launcher_id coin_id inner_puzzlehash target_inner_puzzlehash) release_puzzlehash)
+    (list ASSERT_COIN_ANNOUNCEMENT
+      (sha256 coin_id (sha256 singleton_coin_id nft_launcher_id release_puzzlehash))
+    )
+  )
+
+  (defun compute_nft_lock_conditions (SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH nfts_to_lock offer_nonce conditions_so_far)
+    (if nfts_to_lock
+      (c
+        (offer_assert_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_lock) offer_nonce)
+        (compute_nft_lock_conditions SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (r nfts_to_lock) offer_nonce)
       )
+      conditions_so_far
+    )
   )
 
-  ; assert desired 'wallet' behavior as specified in the offer
-  (defun assert_offer_wallet_asserts (offer_wallet_asserts conditions_so_far)
-    (if (l (r offer_wallet_asserts))
-      (assert_offer_wallet_asserts (r offer_wallet_asserts)
+  (defun compute_nft_unlock_conditions (SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH OFFER_MOD_HASH nfts_to_unlock singleton_coin_id offer_nonce conditions_so_far)
+    (if nfts_to_unlock
+      (c
+        (assert_offer_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) offer_nonce)
         (c
-          (list ASSERT_PUZZLE_ANNOUNCEMENT
-              (sha256
-                  (f (f offer_wallet_asserts))
-                  (r (f offer_wallet_asserts))
-              )
-          ) 
-          conditions_so_far
+          (assert_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) OFFER_MOD_HASH)
+          (c
+            (announce_p2_condition SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH (f nfts_to_unlock) singleton_coin_id OFFER_MOD_HASH)
+            (compute_nft_unlock_conditions SINGLETON_MOD_HASH LAUNCHER_PUZZLE_HASH OFFER_MOD_HASH (r nfts_to_unlock) singleton_coin_id offer_nonce)
```

Comment:
I'm a little baffled as to why this function takes a conditions_so_far but doesn't handle it in the recursive call. Can you help explain @greimela ?